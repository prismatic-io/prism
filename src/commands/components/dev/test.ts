import { promisify } from "util";
import dotenv from "dotenv";
import { Command, Flags, CliUx } from "@oclif/core";
import inquirer, { DistinctQuestion } from "inquirer";
import { snakeCase, upperCase, kebabCase } from "lodash";
import { serverTypes } from "@prismatic-io/spectral"; // FIXME: Get rid of this and stop exporting it in Spectral.
import {
  loadEntrypoint,
  createComponentPackage,
  publishDefinition,
  uploadConnectionIcons,
  uploadFile,
  validateDefinition,
  checkPackageSignature,
} from "../../../utils/component/publish";
import { displayLogs } from "../../../utils/execution/logs";
import {
  buildComponentTestHarnessIntegration,
  componentTestIntegrationName,
  ComponentTestInfo,
} from "../../../utils/integration/definition";
import { importDefinition } from "../../../utils/integration/import";
import {
  deleteIntegration,
  runIntegrationFlow,
} from "../../../utils/integration/invoke";
import { pollForActiveConfigVarState } from "../../../utils/integration/query";
import { Expression } from "../../../utils/integration/export";
import { exists } from "../../../fs";
import { whoAmI } from "../../../utils/user/query";
import { spawnProcess } from "../../../utils/process";
import {
  printFinalStepResults,
  writeFinalStepResults,
} from "../../../utils/execution/stepResults";
import { deleteComponentByKey } from "../../../utils/component/deleteByKey";

const setTimeoutPromise = promisify(setTimeout);

const envVarCase = (name: string): string =>
  upperCase(snakeCase(name)).replace(/\s+/g, "_");

const toInquirerInputType = (
  type: string,
  collection: string | undefined
): Required<DistinctQuestion["type"]> => {
  if (collection) {
    // FIXME: Improve prompting instead of instantly bailing to editor.
    return "editor";
  }

  switch (type) {
    case "boolean":
      return "checkbox";
    case "password":
      return "password";
    case "code":
      return "editor";
    default:
      return "input";
  }
};

const getInputQuestion = ({
  key,
  label,
  type,
  collection,
  default: defaultValue,
}: serverTypes.Input): DistinctQuestion => ({
  type: toInquirerInputType(type, collection),
  name: key,
  message: `${label}:`,
  when: (answers) => {
    const envVar = envVarCase(key);
    const exists = envVar in process.env;
    if (exists) {
      const value = process.env[envVar] ?? "";
      answers[key] = collection
        ? { type: "complex", value: JSON.parse(value) }
        : { type: "value", value };
    }
    return !exists;
  },
  filter: (value) => {
    if (type === "connection") {
      return { type: "configVar", value };
    }

    if (collection) {
      return {
        type: "complex",
        value: JSON.parse(value),
      };
    }

    return {
      type: "value",
      value,
    };
  },
  default: () => (type === "connection" ? "testConnection" : defaultValue),
});

interface PromptAnswers {
  action: serverTypes.Action;
  actionInputs: Record<string, Expression>;
  connection?: serverTypes.Connection;
  connectionInputs?: Record<string, Expression>;
}

const valuesFromAnswers = ({
  action,
  actionInputs,
  connection,
  connectionInputs = {},
}: PromptAnswers): Pick<ComponentTestInfo, "actionInfo" | "connectionInfo"> => {
  const actionValues = Object.entries(actionInputs).reduce<
    ComponentTestInfo["actionInfo"]["values"]
  >(
    (result, [key, value]) => ({
      ...result,
      [key]: value,
    }),
    {}
  );

  const connectionValues = Object.entries(connectionInputs).reduce<
    Required<ComponentTestInfo>["connectionInfo"]["values"]
  >(
    (result, [key, value]) => ({
      ...result,
      [key]: value,
    }),
    {}
  );

  const connectionInfo = connection
    ? { key: connection.key, values: connectionValues }
    : undefined;

  return {
    actionInfo: { key: action.key, values: actionValues },
    connectionInfo,
  };
};

export default class TestCommand extends Command {
  static description =
    "Run an action of a component within a test integration in the integration runner";
  static flags = {
    envPath: Flags.string({
      required: false,
      default: ".env",
      char: "e",
      description: "Path to dotenv file to load for supplying testing values",
    }),
    build: Flags.boolean({
      required: false,
      default: true,
      allowNo: true,
      char: "b",
      description: "Build the component prior to testing",
    }),
    "output-file": Flags.string({
      required: false,
      char: "o",
      description: "Output the results of the action to a specified file",
    }),
    "print-results": Flags.boolean({
      required: false,
      default: false,
      description: "Print the results of the action to stdout",
    }),
    "clean-up": Flags.boolean({
      required: false,
      default: false,
      description:
        "Clean up the integration and temporary component after running the action",
    }),
  };

  async run() {
    const {
      flags: {
        envPath,
        build,
        "output-file": outputFile,
        "print-results": printResults,
        "clean-up": cleanUp,
      },
    } = await this.parse(TestCommand);

    // Save the current working directory, so we can return later after moving to dist/
    const cwd = process.cwd();

    if (build) {
      console.log("Building component...");
      await spawnProcess(["npm", "run", "build"], {});
    }

    if (await exists(envPath)) {
      const { error } = dotenv.config({ path: envPath, override: true });
      if (error) {
        CliUx.ux.error(`Failed to load specified dotenv file: ${error}`, {
          exit: 1,
        });
      }
    }

    CliUx.ux.action.start("Validating Component");

    const { name } = await whoAmI();
    if (!name) {
      CliUx.ux.error(
        "Failed to determine the name of the currently logged in user.",
        {
          exit: 1,
        }
      );
    }

    const testingKey = kebabCase(name);

    const definition = await loadEntrypoint();
    const { key: componentKey, public: isPublic } = definition;
    definition.key = `${componentKey}-${testingKey}-testing`;
    definition.display.label = `${definition.display.label} ${name} Testing`;
    await validateDefinition(definition);

    const packagePath = await createComponentPackage();
    const signatureMatches = await checkPackageSignature(
      definition,
      packagePath
    );

    CliUx.ux.action.stop();

    if (!signatureMatches) {
      CliUx.ux.action.start("Publishing Component");

      const { iconUploadUrl, packageUploadUrl, connectionIconUploadUrls } =
        await publishDefinition(definition);

      const {
        display: { iconPath },
      } = definition;
      await uploadFile(iconPath, iconUploadUrl);
      await uploadConnectionIcons(definition, connectionIconUploadUrls);
      await uploadFile(packagePath, packageUploadUrl);

      CliUx.ux.action.stop();
    }

    const publishedTimestamp = Date.now();

    const actions = definition.actions || {};

    const { action } = await inquirer.prompt<{ action: serverTypes.Action }>({
      type: "list",
      name: "action",
      message: "Action:",
      choices: Object.entries(actions).map(
        ([
          key,
          {
            display: { label },
          },
        ]) => ({
          name: label,
          value: key,
          short: key,
        })
      ),
      default: Object.keys(actions)[0],
      filter: (value: string) => actions[value],
    });

    const { inputs } = action;

    // Ask for values of action's inputs
    const actionInputs = await inquirer.prompt(
      inputs.map((i) => getInputQuestion(i))
    );

    const answers = {
      action,
      actionInputs,
    };

    // Ask about Connection to test if there is a connection type input
    const hasConnection = inputs.some(({ type }) => type === "connection");
    if (hasConnection) {
      const connections = definition.connections || [];
      const { connection } = await inquirer.prompt<{
        connection: serverTypes.Connection;
      }>({
        type: "list",
        name: "connection",
        message: "Connection:",
        choices: connections.map(({ key, label }) => ({
          name: label,
          value: key,
          short: key,
        })),
        default: Object.keys(connections)[0],
        filter: (value) => {
          const [connection] = connections.filter(({ key }) => value === key);
          return connection;
        },
      });

      // Prompt for connection's inputs
      const { inputs } = connection;
      const connectionInputs = await inquirer.prompt(
        inputs
          .filter(({ shown }) => shown === undefined || shown === true)
          .map((i) => getInputQuestion(i))
      );

      Object.assign(answers, { connection, connectionInputs });
    }

    const { actionInfo, connectionInfo } = valuesFromAnswers(answers);

    CliUx.ux.action.start("Assembling test integration");

    // FIXME: Wait for version to be available but due to issues we have to do a static wait.
    const wait = 5000 - (Date.now() - publishedTimestamp);
    if (wait > 0) {
      await setTimeoutPromise(wait);
    }

    // Build up YAML
    const harnessYaml = await buildComponentTestHarnessIntegration({
      integrationInfo: {
        name: componentTestIntegrationName(componentKey, name),
      },
      componentInfo: { key: definition.key, isPublic: isPublic ?? false },
      actionInfo,
      connectionInfo,
    });

    CliUx.ux.action.stop();
    CliUx.ux.action.start("Updating test Integration");

    const {
      integrationId,
      flows: [{ id: flowId }],
      pendingAuthorizations,
    } = await importDefinition(harnessYaml);

    CliUx.ux.action.stop();

    // Prompt for user authorization of pending connections
    if (pendingAuthorizations.length > 0) {
      const [{ id, url }] = pendingAuthorizations;
      if (!url) {
        throw new Error(
          "Did not receive a valid URL for authorization. Verify your Connection inputs."
        );
      }

      CliUx.ux.url("Authorize URL", url);
      await CliUx.ux.anykey(
        "Press any key to open your browser and authorize the Connection"
      );
      await CliUx.ux.open(url);

      CliUx.ux.action.start("Waiting for Connection authorization");

      await pollForActiveConfigVarState(integrationId, id);

      CliUx.ux.action.stop();
    }

    CliUx.ux.action.start("Running test Integration");

    const { executionId } = await runIntegrationFlow({ integrationId, flowId });

    await displayLogs(executionId);

    if (outputFile) {
      process.chdir(cwd);
      console.log(`Writing step results to ${outputFile}`);
      await writeFinalStepResults(executionId, outputFile);
    }

    if (printResults) {
      await printFinalStepResults(executionId);
    }

    if (cleanUp) {
      CliUx.ux.action.start(`Cleaning up test Integration (${integrationId})`);
      await deleteIntegration(integrationId);
      CliUx.ux.action.stop();

      CliUx.ux.action.start(`Cleaning up test component (${definition.key})`);
      await deleteComponentByKey(definition.key);
      CliUx.ux.action.stop();
    }

    CliUx.ux.action.stop();
  }
}
