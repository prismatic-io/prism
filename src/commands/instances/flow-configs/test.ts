import { Command, Flags, Args, ux } from "@oclif/core";
import { gqlRequest, gql } from "../../../graphql";

interface LogNode {
  [index: string]: unknown;
  timestamp: string;
  severity: string;
  message: string;
}

interface FetchLogsResult {
  logs: LogNode[];
  cursor: string | undefined;
  executionComplete: boolean | undefined;
}

export default class TestCommand extends Command {
  static description = "Test a Flow Config of an Instance";
  static args = {
    flowConfig: Args.string({
      description: "ID of a Flow Config to test",
      required: true,
    }),
  };

  static flags = {
    ...ux.table.flags({ only: ["extended", "columns"] }),
    tail: Flags.boolean({
      required: false,
      char: "t",
      description: "Tail logs of the flow config test run",
    }),
    payload: Flags.string({
      required: false,
      char: "p",
      description:
        "Optional JSON-formatted data payload to submit with the test",
    }),
    contentType: Flags.string({
      required: false,
      char: "c",
      description: "Optional content-type for the test payload",
    }),
  };

  async run() {
    const {
      args: { flowConfig },
      flags: { tail, payload, contentType },
    } = await this.parse(TestCommand);

    const result = await gqlRequest({
      document: gql`
        mutation testInstanceFlowConfig(
          $id: ID!
          $payload: String
          $contentType: String
        ) {
          testInstanceFlowConfig(
            input: { id: $id, payload: $payload, contentType: $contentType }
          ) {
            testInstanceFlowConfigResult {
              flowConfig {
                id
              }
              execution {
                id
              }
            }
            errors {
              field
              messages
            }
          }
        }
      `,
      variables: {
        id: flowConfig,
        payload,
        contentType,
      },
    });

    if (!tail) {
      return;
    }

    const executionId =
      result.testInstanceFlowConfig.testInstanceFlowConfigResult.execution.id;
    await this.tailLogs(executionId);
  }

  private async tailLogs(executionId: string) {
    const { flags } = await this.parse(TestCommand);

    let nextCursor: string | undefined = undefined;
    // eslint-disable-next-line no-constant-condition
    while (true) {
      await ux.wait(500);

      const result: any = await this.fetchLogs(executionId, nextCursor);
      if (result === undefined) continue;

      const { logs, cursor, executionComplete } = result;

      nextCursor = cursor;

      ux.table(
        logs,
        {
          timestamp: {},
          severity: {
            minWidth: 12,
          },
          message: {},
        },
        { ...flags, "no-header": true }
      );

      if (executionComplete) return;
    }
  }

  private async fetchLogs(
    executionId: string,
    nextCursor?: string
  ): Promise<FetchLogsResult | undefined> {
    const results = await gqlRequest({
      document: gql`
        query listInstanceTestLogs($executionId: ID!, $nextCursor: String) {
          logs(
            executionResult: $executionId
            after: $nextCursor
            orderBy: { field: TIMESTAMP, direction: ASC }
          ) {
            edges {
              node {
                timestamp
                severity
                message
              }
              cursor
            }
          }
        }
      `,
      variables: {
        executionId,
        nextCursor,
      },
    });

    const {
      edges,
    }: {
      edges: { cursor?: string; node: LogNode }[];
    } = results.logs;
    if (!edges || edges.length === 0) {
      return undefined;
    }

    const logs = edges.map(({ node }) => node);
    const executionComplete = logs.reduce<boolean>(
      (result: boolean, { message }) =>
        result || message.startsWith("Ending Instance Execution"),
      false
    );

    const { cursor } = edges[edges.length - 1];
    return { logs, cursor, executionComplete };
  }
}
