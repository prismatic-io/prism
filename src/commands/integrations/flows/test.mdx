```bash
# Test an integration flow with a payload file and tail the logs and step results
prism integrations:flows:test \
    -p=some_payload_file.xml
    -c=application/xml
    --tail-logs
    --tail-results
```
