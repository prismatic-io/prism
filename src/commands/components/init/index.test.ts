import { expect, it, describe } from "bun:test";
import fs from "fs";
import path from "path";
import InitializeComponent from ".";

const tempPath = "src/commands/components/init/temp";
if (!fs.existsSync(tempPath)) {
  fs.mkdirSync(tempPath);
}
process.chdir(tempPath);

const componentGenerationTimeout = 5 * 60 * 1000; // 5 minutes

interface SpecMeta {
  specName: string;
  componentName: string;
}

const specFiles = fs.readdirSync("../fixtures/specs");
const specs = specFiles.reduce(
  (result, fileName) => {
    if (fileName.endsWith(".wsdl")) {
      const [componentName] = fileName.toLowerCase().split(".");
      result.wsdl.push({
        specName: fileName,
        componentName,
      });
    }
    return result;
  },
  { wsdl: [] as SpecMeta[] },
);

const generatedFiles = {
  wsdl: ["actions.ts", "triggers.ts", "index.ts", "inputs.ts"],
};

// NOTE: OpenAPI and basic scaffolding tests are moved to the spectral repo under the generator package
describe("component generation tests", () => {
  describe("wsdl", () => {
    for (const { specName, componentName } of specs.wsdl) {
      describe(`${specName} generation`, () => {
        it(
          "should generate successfully",
          async () => {
            await InitializeComponent.run([
              componentName,
              `--wsdl-path=../fixtures/specs/${specName}`,
            ]);
          },
          componentGenerationTimeout,
        );

        for (const filename of generatedFiles.wsdl) {
          it(`should match ${filename} snapshot`, async () => {
            const contents = fs.readFileSync(path.join(componentName, "src", filename)).toString();
            expect(contents).toMatchSnapshot(`${componentName}-${filename}`);
          });
        }
      });
    }
  });
});
