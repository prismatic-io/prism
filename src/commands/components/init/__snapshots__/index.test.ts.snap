// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`component generation tests wsdl CountryInfoService.wsdl generation should match actions.ts snapshot: countryinfoservice-actions.ts 1`] = `
"import { action } from "@prismatic-io/spectral";
import * as path from "path";
import { createClientAsync } from "../countryinfoservice/index";
import {
  sCountryISOCode,
  sCountryName,
  sCurrencyISOCode,
  sISOCode,
  sISOCurrencyCode,
  sLanguageName,
} from "./inputs";
export const capitalCityAsync = action({
  display: {
    label: "Capital City Async",
    description: \`Generated by Prismatic\`,
  },
  perform: async (
    { logger },
    { sCountryISOCode, headers }: Record<string, any>
  ) => {
    const client = await createClientAsync(
      path.join(__dirname, "countryInfoService.wsdl"),
      headers
    );
    const [result] = await client.CapitalCityAsync({ sCountryISOCode });
    return { data: result };
  },
  inputs: {
    sCountryISOCode,
    headers: {
      label: "Headers",
      type: "code",
      comments: "Provide headers to the SOAP client",
      default: "{}",
    },
  },
});
export const countriesUsingCurrencyAsync = action({
  display: {
    label: "Countries Using Currency Async",
    description: \`Generated by Prismatic\`,
  },
  perform: async (
    { logger },
    { sISOCurrencyCode, headers }: Record<string, any>
  ) => {
    const client = await createClientAsync(
      path.join(__dirname, "countryInfoService.wsdl"),
      headers
    );
    const [result] = await client.CountriesUsingCurrencyAsync({
      sISOCurrencyCode,
    });
    return { data: result };
  },
  inputs: {
    sISOCurrencyCode,
    headers: {
      label: "Headers",
      type: "code",
      comments: "Provide headers to the SOAP client",
      default: "{}",
    },
  },
});
export const countryCurrencyAsync = action({
  display: {
    label: "Country Currency Async",
    description: \`Generated by Prismatic\`,
  },
  perform: async (
    { logger },
    { sCountryISOCode, headers }: Record<string, any>
  ) => {
    const client = await createClientAsync(
      path.join(__dirname, "countryInfoService.wsdl"),
      headers
    );
    const [result] = await client.CountryCurrencyAsync({ sCountryISOCode });
    return { data: result };
  },
  inputs: {
    sCountryISOCode,
    headers: {
      label: "Headers",
      type: "code",
      comments: "Provide headers to the SOAP client",
      default: "{}",
    },
  },
});
export const countryFlagAsync = action({
  display: {
    label: "Country Flag Async",
    description: \`Generated by Prismatic\`,
  },
  perform: async (
    { logger },
    { sCountryISOCode, headers }: Record<string, any>
  ) => {
    const client = await createClientAsync(
      path.join(__dirname, "countryInfoService.wsdl"),
      headers
    );
    const [result] = await client.CountryFlagAsync({ sCountryISOCode });
    return { data: result };
  },
  inputs: {
    sCountryISOCode,
    headers: {
      label: "Headers",
      type: "code",
      comments: "Provide headers to the SOAP client",
      default: "{}",
    },
  },
});
export const countryIntPhoneCodeAsync = action({
  display: {
    label: "Country Int Phone Code Async",
    description: \`Generated by Prismatic\`,
  },
  perform: async (
    { logger },
    { sCountryISOCode, headers }: Record<string, any>
  ) => {
    const client = await createClientAsync(
      path.join(__dirname, "countryInfoService.wsdl"),
      headers
    );
    const [result] = await client.CountryIntPhoneCodeAsync({ sCountryISOCode });
    return { data: result };
  },
  inputs: {
    sCountryISOCode,
    headers: {
      label: "Headers",
      type: "code",
      comments: "Provide headers to the SOAP client",
      default: "{}",
    },
  },
});
export const countryIsoCodeAsync = action({
  display: {
    label: "Country Iso Code Async",
    description: \`Generated by Prismatic\`,
  },
  perform: async (
    { logger },
    { sCountryName, headers }: Record<string, any>
  ) => {
    const client = await createClientAsync(
      path.join(__dirname, "countryInfoService.wsdl"),
      headers
    );
    const [result] = await client.CountryISOCodeAsync({ sCountryName });
    return { data: result };
  },
  inputs: {
    sCountryName,
    headers: {
      label: "Headers",
      type: "code",
      comments: "Provide headers to the SOAP client",
      default: "{}",
    },
  },
});
export const countryNameAsync = action({
  display: {
    label: "Country Name Async",
    description: \`Generated by Prismatic\`,
  },
  perform: async (
    { logger },
    { sCountryISOCode, headers }: Record<string, any>
  ) => {
    const client = await createClientAsync(
      path.join(__dirname, "countryInfoService.wsdl"),
      headers
    );
    const [result] = await client.CountryNameAsync({ sCountryISOCode });
    return { data: result };
  },
  inputs: {
    sCountryISOCode,
    headers: {
      label: "Headers",
      type: "code",
      comments: "Provide headers to the SOAP client",
      default: "{}",
    },
  },
});
export const currencyNameAsync = action({
  display: {
    label: "Currency Name Async",
    description: \`Generated by Prismatic\`,
  },
  perform: async (
    { logger },
    { sCurrencyISOCode, headers }: Record<string, any>
  ) => {
    const client = await createClientAsync(
      path.join(__dirname, "countryInfoService.wsdl"),
      headers
    );
    const [result] = await client.CurrencyNameAsync({ sCurrencyISOCode });
    return { data: result };
  },
  inputs: {
    sCurrencyISOCode,
    headers: {
      label: "Headers",
      type: "code",
      comments: "Provide headers to the SOAP client",
      default: "{}",
    },
  },
});
export const fullCountryInfoAllCountriesAsync = action({
  display: {
    label: "Full Country Info All Countries Async",
    description: \`Generated by Prismatic\`,
  },
  perform: async ({ logger }, { headers }: Record<string, any>) => {
    const client = await createClientAsync(
      path.join(__dirname, "countryInfoService.wsdl"),
      headers
    );
    const [result] = await client.FullCountryInfoAllCountriesAsync({});
    return { data: result };
  },
  inputs: {
    headers: {
      label: "Headers",
      type: "code",
      comments: "Provide headers to the SOAP client",
      default: "{}",
    },
  },
});
export const fullCountryInfoAsync = action({
  display: {
    label: "Full Country Info Async",
    description: \`Generated by Prismatic\`,
  },
  perform: async (
    { logger },
    { sCountryISOCode, headers }: Record<string, any>
  ) => {
    const client = await createClientAsync(
      path.join(__dirname, "countryInfoService.wsdl"),
      headers
    );
    const [result] = await client.FullCountryInfoAsync({ sCountryISOCode });
    return { data: result };
  },
  inputs: {
    sCountryISOCode,
    headers: {
      label: "Headers",
      type: "code",
      comments: "Provide headers to the SOAP client",
      default: "{}",
    },
  },
});
export const languageIsoCodeAsync = action({
  display: {
    label: "Language Iso Code Async",
    description: \`Generated by Prismatic\`,
  },
  perform: async (
    { logger },
    { sLanguageName, headers }: Record<string, any>
  ) => {
    const client = await createClientAsync(
      path.join(__dirname, "countryInfoService.wsdl"),
      headers
    );
    const [result] = await client.LanguageISOCodeAsync({ sLanguageName });
    return { data: result };
  },
  inputs: {
    sLanguageName,
    headers: {
      label: "Headers",
      type: "code",
      comments: "Provide headers to the SOAP client",
      default: "{}",
    },
  },
});
export const languageNameAsync = action({
  display: {
    label: "Language Name Async",
    description: \`Generated by Prismatic\`,
  },
  perform: async ({ logger }, { sISOCode, headers }: Record<string, any>) => {
    const client = await createClientAsync(
      path.join(__dirname, "countryInfoService.wsdl"),
      headers
    );
    const [result] = await client.LanguageNameAsync({ sISOCode });
    return { data: result };
  },
  inputs: {
    sISOCode,
    headers: {
      label: "Headers",
      type: "code",
      comments: "Provide headers to the SOAP client",
      default: "{}",
    },
  },
});
export const listOfContinentsByCodeAsync = action({
  display: {
    label: "List Of Continents By Code Async",
    description: \`Generated by Prismatic\`,
  },
  perform: async ({ logger }, { headers }: Record<string, any>) => {
    const client = await createClientAsync(
      path.join(__dirname, "countryInfoService.wsdl"),
      headers
    );
    const [result] = await client.ListOfContinentsByCodeAsync({});
    return { data: result };
  },
  inputs: {
    headers: {
      label: "Headers",
      type: "code",
      comments: "Provide headers to the SOAP client",
      default: "{}",
    },
  },
});
export const listOfContinentsByNameAsync = action({
  display: {
    label: "List Of Continents By Name Async",
    description: \`Generated by Prismatic\`,
  },
  perform: async ({ logger }, { headers }: Record<string, any>) => {
    const client = await createClientAsync(
      path.join(__dirname, "countryInfoService.wsdl"),
      headers
    );
    const [result] = await client.ListOfContinentsByNameAsync({});
    return { data: result };
  },
  inputs: {
    headers: {
      label: "Headers",
      type: "code",
      comments: "Provide headers to the SOAP client",
      default: "{}",
    },
  },
});
export const listOfCountryNamesByCodeAsync = action({
  display: {
    label: "List Of Country Names By Code Async",
    description: \`Generated by Prismatic\`,
  },
  perform: async ({ logger }, { headers }: Record<string, any>) => {
    const client = await createClientAsync(
      path.join(__dirname, "countryInfoService.wsdl"),
      headers
    );
    const [result] = await client.ListOfCountryNamesByCodeAsync({});
    return { data: result };
  },
  inputs: {
    headers: {
      label: "Headers",
      type: "code",
      comments: "Provide headers to the SOAP client",
      default: "{}",
    },
  },
});
export const listOfCountryNamesByNameAsync = action({
  display: {
    label: "List Of Country Names By Name Async",
    description: \`Generated by Prismatic\`,
  },
  perform: async ({ logger }, { headers }: Record<string, any>) => {
    const client = await createClientAsync(
      path.join(__dirname, "countryInfoService.wsdl"),
      headers
    );
    const [result] = await client.ListOfCountryNamesByNameAsync({});
    return { data: result };
  },
  inputs: {
    headers: {
      label: "Headers",
      type: "code",
      comments: "Provide headers to the SOAP client",
      default: "{}",
    },
  },
});
export const listOfCountryNamesGroupedByContinentAsync = action({
  display: {
    label: "List Of Country Names Grouped By Continent Async",
    description: \`Generated by Prismatic\`,
  },
  perform: async ({ logger }, { headers }: Record<string, any>) => {
    const client = await createClientAsync(
      path.join(__dirname, "countryInfoService.wsdl"),
      headers
    );
    const [result] = await client.ListOfCountryNamesGroupedByContinentAsync({});
    return { data: result };
  },
  inputs: {
    headers: {
      label: "Headers",
      type: "code",
      comments: "Provide headers to the SOAP client",
      default: "{}",
    },
  },
});
export const listOfCurrenciesByCodeAsync = action({
  display: {
    label: "List Of Currencies By Code Async",
    description: \`Generated by Prismatic\`,
  },
  perform: async ({ logger }, { headers }: Record<string, any>) => {
    const client = await createClientAsync(
      path.join(__dirname, "countryInfoService.wsdl"),
      headers
    );
    const [result] = await client.ListOfCurrenciesByCodeAsync({});
    return { data: result };
  },
  inputs: {
    headers: {
      label: "Headers",
      type: "code",
      comments: "Provide headers to the SOAP client",
      default: "{}",
    },
  },
});
export const listOfCurrenciesByNameAsync = action({
  display: {
    label: "List Of Currencies By Name Async",
    description: \`Generated by Prismatic\`,
  },
  perform: async ({ logger }, { headers }: Record<string, any>) => {
    const client = await createClientAsync(
      path.join(__dirname, "countryInfoService.wsdl"),
      headers
    );
    const [result] = await client.ListOfCurrenciesByNameAsync({});
    return { data: result };
  },
  inputs: {
    headers: {
      label: "Headers",
      type: "code",
      comments: "Provide headers to the SOAP client",
      default: "{}",
    },
  },
});
export const listOfLanguagesByCodeAsync = action({
  display: {
    label: "List Of Languages By Code Async",
    description: \`Generated by Prismatic\`,
  },
  perform: async ({ logger }, { headers }: Record<string, any>) => {
    const client = await createClientAsync(
      path.join(__dirname, "countryInfoService.wsdl"),
      headers
    );
    const [result] = await client.ListOfLanguagesByCodeAsync({});
    return { data: result };
  },
  inputs: {
    headers: {
      label: "Headers",
      type: "code",
      comments: "Provide headers to the SOAP client",
      default: "{}",
    },
  },
});
export const listOfLanguagesByNameAsync = action({
  display: {
    label: "List Of Languages By Name Async",
    description: \`Generated by Prismatic\`,
  },
  perform: async ({ logger }, { headers }: Record<string, any>) => {
    const client = await createClientAsync(
      path.join(__dirname, "countryInfoService.wsdl"),
      headers
    );
    const [result] = await client.ListOfLanguagesByNameAsync({});
    return { data: result };
  },
  inputs: {
    headers: {
      label: "Headers",
      type: "code",
      comments: "Provide headers to the SOAP client",
      default: "{}",
    },
  },
});

export default {
  capitalCityAsync,
  countriesUsingCurrencyAsync,
  countryCurrencyAsync,
  countryFlagAsync,
  countryIntPhoneCodeAsync,
  countryIsoCodeAsync,
  countryNameAsync,
  currencyNameAsync,
  fullCountryInfoAllCountriesAsync,
  fullCountryInfoAsync,
  languageIsoCodeAsync,
  languageNameAsync,
  listOfContinentsByCodeAsync,
  listOfContinentsByNameAsync,
  listOfCountryNamesByCodeAsync,
  listOfCountryNamesByNameAsync,
  listOfCountryNamesGroupedByContinentAsync,
  listOfCurrenciesByCodeAsync,
  listOfCurrenciesByNameAsync,
  listOfLanguagesByCodeAsync,
  listOfLanguagesByNameAsync,
};
"
`;

exports[`component generation tests wsdl CountryInfoService.wsdl generation should match index.ts snapshot: countryinfoservice-index.ts 1`] = `
"import { component } from "@prismatic-io/spectral";
import actions from "./actions";
import triggers from "./triggers";
import dataSources from "./dataSources";
import connections from "./connections";

export default component({
  key: "countryinfoservice",
  public: false,
  display: {
    label: "countryinfoservice",
    description: "Prism-generated Component",
    iconPath: "icon.png",
  },
  actions,
  triggers,
  dataSources,
  connections,
});
"
`;

exports[`component generation tests wsdl CountryInfoService.wsdl generation should match inputs.ts snapshot: countryinfoservice-inputs.ts 1`] = `
"import { input } from "@prismatic-io/spectral";
export const sCountryISOCode = input({
  label: "SCountryIsoCode",
  type: "data",
  required: true,
});
export const sCountryName = input({
  label: "SCountryName",
  type: "data",
  required: true,
});
export const sCurrencyISOCode = input({
  label: "SCurrencyIsoCode",
  type: "data",
  required: true,
});
export const sISOCode = input({
  label: "SIsoCode",
  type: "data",
  required: true,
});
export const sISOCurrencyCode = input({
  label: "SIsoCurrencyCode",
  type: "data",
  required: true,
});
export const sLanguageName = input({
  label: "SLanguageName",
  type: "data",
  required: true,
});
"
`;

exports[`component generation tests wsdl CountryInfoService.wsdl generation should match triggers.ts snapshot: countryinfoservice-triggers.ts 1`] = `
"import { trigger } from "@prismatic-io/spectral";

export const myTrigger = trigger({
  display: {
    label: "My Trigger",
    description: "This is my trigger",
  },
  perform: async (context, payload, params) => {
    console.log("My Trigger params", params);
    return Promise.resolve({
      payload,
    });
  },
  inputs: {},
  synchronousResponseSupport: "valid",
  scheduleSupport: "valid",
});

export default { myTrigger };
"
`;

exports[`component generation tests wsdl NumberConversion.wsdl generation should match actions.ts snapshot: numberconversion-actions.ts 1`] = `
"import { action } from "@prismatic-io/spectral";
import * as path from "path";
import { createClientAsync } from "../numberconversion/index";
import { dNum, ubiNum } from "./inputs";
export const numberToDollarsAsync = action({
  display: {
    label: "Number To Dollars Async",
    description: \`Generated by Prismatic\`,
  },
  perform: async ({ logger }, { dNum, headers }: Record<string, any>) => {
    const client = await createClientAsync(
      path.join(__dirname, "numberConversion.wsdl"),
      headers
    );
    const [result] = await client.NumberToDollarsAsync({ dNum });
    return { data: result };
  },
  inputs: {
    dNum,
    headers: {
      label: "Headers",
      type: "code",
      comments: "Provide headers to the SOAP client",
      default: "{}",
    },
  },
});
export const numberToWordsAsync = action({
  display: {
    label: "Number To Words Async",
    description: \`Generated by Prismatic\`,
  },
  perform: async ({ logger }, { ubiNum, headers }: Record<string, any>) => {
    const client = await createClientAsync(
      path.join(__dirname, "numberConversion.wsdl"),
      headers
    );
    const [result] = await client.NumberToWordsAsync({ ubiNum });
    return { data: result };
  },
  inputs: {
    ubiNum,
    headers: {
      label: "Headers",
      type: "code",
      comments: "Provide headers to the SOAP client",
      default: "{}",
    },
  },
});

export default {
  numberToDollarsAsync,
  numberToWordsAsync,
};
"
`;

exports[`component generation tests wsdl NumberConversion.wsdl generation should match index.ts snapshot: numberconversion-index.ts 1`] = `
"import { component } from "@prismatic-io/spectral";
import actions from "./actions";
import triggers from "./triggers";
import dataSources from "./dataSources";
import connections from "./connections";

export default component({
  key: "numberconversion",
  public: false,
  display: {
    label: "numberconversion",
    description: "Prism-generated Component",
    iconPath: "icon.png",
  },
  actions,
  triggers,
  dataSources,
  connections,
});
"
`;

exports[`component generation tests wsdl NumberConversion.wsdl generation should match inputs.ts snapshot: numberconversion-inputs.ts 1`] = `
"import { input } from "@prismatic-io/spectral";
export const dNum = input({
  label: "DNum",
  type: "data",
  required: true,
});
export const ubiNum = input({
  label: "UbiNum",
  type: "data",
  required: true,
});
"
`;

exports[`component generation tests wsdl NumberConversion.wsdl generation should match triggers.ts snapshot: numberconversion-triggers.ts 1`] = `
"import { trigger } from "@prismatic-io/spectral";

export const myTrigger = trigger({
  display: {
    label: "My Trigger",
    description: "This is my trigger",
  },
  perform: async (context, payload, params) => {
    console.log("My Trigger params", params);
    return Promise.resolve({
      payload,
    });
  },
  inputs: {},
  synchronousResponseSupport: "valid",
  scheduleSupport: "valid",
});

export default { myTrigger };
"
`;
